src/main/java/net/fabricmc/loader/impl/launch/knot/MixinServiceKnotBootstrap.java

```diff
@@ -18,7 +18,17 @@
 
 import org.spongepowered.asm.service.IMixinServiceBootstrap;
 
+import net.fabricmc.loader.impl.launch.FabricLauncherBase;
+
 public class MixinServiceKnotBootstrap implements IMixinServiceBootstrap {
+	public MixinServiceKnotBootstrap() {
+		if (!FabricLauncherBase.getLauncher().useFabricMixinServices()) {
+			// If an exception is thrown here, Mixin is designed to skip over this service.
```

Comment:
Isn't the `false` return in `MixinServiceKnot.isValid` the way to prevent mixin from loading the service? Throwing an exception is costly and isn't a good way to prevent loading of service, as you have no way to distinguish it from other exceptions like ones generated from adding to an immutable list.

Reply:
Mixin itself does this in both the [ModLauncher](https://github.com/FabricMC/Mixin/blob/921353610dfe0a2860c89262f306b79b8b95c199/src/modlauncher/java/org/spongepowered/asm/service/modlauncher/MixinServiceModLauncherBootstrap.java#L49-L53) and [LaunchWrapper](https://github.com/FabricMC/Mixin/blob/921353610dfe0a2860c89262f306b79b8b95c199/src/launchwrapper/java/org/spongepowered/asm/service/mojang/MixinServiceLaunchWrapperBootstrap.java#L58-L62) implementations, albeit in `bootstrap` instead of the constructor and throwing a more precise exception (I somehow missed both of these details, will need to fix).